src/drivers/clock/sp804_timer/sp804_timer.c

```diff
@@ -0,0 +1,104 @@
+/**
+ * @file sp804_timer.c
+ * @brief ARM Dual-Timer (SP804) Module driver
+ * @author Nishant Malpani <nish.malpani25@gmail.com>
+ * @date 2020-03-07
+ */
+
+#include <drivers/common/memory.h>
+#include <hal/clock.h>
+#include <hal/reg.h>
+#include <kernel/irq.h>
+#include <kernel/printk.h>
+#include <kernel/time/clock_source.h>
+#include <kernel/time/time_device.h>
+
+#include <framework/mod/options.h>
+#include <embox/unit.h>
+
+#define SP804_BASE      OPTION_GET(NUMBER, base_addr) /* Base address */
+#define SP804_IRQ       OPTION_GET(NUMBER, irq_nr) /* Interrupt Request number */
+#define SP804_TARGET_HZ OPTION_GET(NUMBER, target_freq) /* Frequency to run the timer at */
+#define SP804_TIMCLK    100000000 /* Frequency of the SP804 clock timer - 100 MHz */
+
+#define SP804_T1_LR     (SP804_BASE + 0x00) /* Timer1 Load Register */
+#define SP804_T1_VR     (SP804_BASE + 0x04) /* Timer1 Current Value Register */
+#define SP804_T1_CR     (SP804_BASE + 0x08) /* Timer1 Control Register */
+#define SP804_T1_ICR    (SP804_BASE + 0x0C) /* Timer1 Interrupt Clear Register */
+#define SP804_T1_RIS    (SP804_BASE + 0x10) /* Timer1 Raw Interrupt Status Register */
+#define SP804_T1_MIS    (SP804_BASE + 0x14) /* Timer1 Masked Interrupt Status Register */
+#define SP804_T1_BGL    (SP804_BASE + 0x18) /* Timer1 Background Load Register */
+
+#define SP804_CR_MASK           (0b11101111) /* Mask for bits 0-7 */
+#define SP804_CR_ONESHOT        (0 << 0) /* OneShot bit; 0 -> wrapping mode, 1 -> one-shot mode */
+#define SP804_CR_TIMERSIZE      (0 << 1) /* TimerSize bit; 0 -> 16-bit counter, 1 -> 32-bit counter */
+#define SP804_CR_TIMERPRE       (0 << 2) /* Prescale bits; 0 -> div by 1, 1 -> div by 16, 2 -> div by 256 */
+#define SP804_CR_INTEN          (1 << 5) /* InterruptEnable bit */
+#define SP804_CR_TIMERMODE      (0 << 6) /* TimerMode bit; 0 -> free-running mode, 1 -> periodic mode */
+#define SP804_CR_TIMEREN        (1 << 7) /* TimerEnable bit */
+#define SP804_ICR_CLEAR         (1 << 0) /* InterruptClear bit */
+
+#define SP804_PRESCALE          1 /* Prescaler operation - divide the rate of master clock */
+#define SP804_LR_VAL            ((SP804_TIMCLK) / (SP804_PRESCALE * SP804_TARGET_HZ))
+
+static struct clock_source sp804_clock_source;
+static irq_return_t clock_handler(unsigned int irq_nr, void *data) {
+    clock_tick_handler(irq_nr, data);
```

Comment:
maybe call clock_tick_handler() after clear interrupt and load a new value?

Reply:
I cannot think of any issues the way you've suggested may create for now, but here I've gone along with what other drivers have implemented [1] [2]. You want me to go ahead and still change it and call `clock_tick_handler()` in the end? 

[1] https://github.com/embox/embox/blob/master/src/drivers/clock/cortexa9_timer/cortexa9_timer.c#L48
[2] https://github.com/embox/embox/blob/master/src/drivers/clock/epit/epit.c#L44